[10.0][MIG] l10n_ch_lsv_dd

ecino opened the pull request:
Migrate module l10n_ch_lsv_dd from version 8.0 up to 10.0

The module allows to export payment orders in valid swiss formats (Postfinance + swiss SEPA)

Commit:
Ported the module l10n_ch_lsv_dd from v8 to v9.

- Allow to generate the payment files from the button
  Generate Payment File of a payment order.
- Update tests

Commit:
Implement pain.008.001.02.ch.03

Commit:
Various bug fixes and improvements

- Added some more tests

ecino committed:
[10.0] Migrate l10n_ch_lsv_dd

- Various bug fixes
- Adapt lsv_dd tests to version 10

ecino commented:
The test of the lsv_dd module is no more adapted to version 10, I will have to change that.

yvaucher commented:
@ecino I pass this as Work in progress for the test fixing.

ecino commented:
@yvaucher I fixed the tests, do you have any idea to overcome this missing account error? locally the tests are working, do we miss some configuration for travis pull requests?

yvaucher commented:
@ecino I think it needs the basic demo accouts from account module.

You might want to try with:

``` python
        self._load('account', 'test', 'account_minimal_test.xml')
```

As it was done in l10n_ch_dta module

ecino commented:
Thanks, I will fix this. I need as well to change the free_invoice functionality as the payment_order changed in v10. I'm thinking of splitting this functionnality into a new module that would suit in the bank-payment repository.

ecino commented:
@yvaucher I've updated the pull request but I'm still getting the same error even though I added the load of test data. Any idea?
https://github.com/OCA/l10n-switzerland/pull/362/files#diff-395c3f0248b511c55c0ea42f48f0e2f6R39

ecino commented:
@yvaucher Finally I managed to fix these tests! The module is now ready to be merged.

ecino committed:
CO-1571 FIX lsv file generation

yvaucher commented on l10n_ch_lsv_dd/README.rst:
```diff
@@ -0,0 +1,106 @@
+
+To install this module, you need to:
+
+* do this ...
```
Here you probably want to state any dependency or that installation has no special requirements.

yvaucher commented on l10n_ch_lsv_dd/README.rst:
```diff
@@ -0,0 +1,106 @@
+   :target: https://runbot.odoo-community.org/runbot/125/8.0
+
+.. repo_id is available in https://github.com/OCA/maintainer-tools/blob/master/tools/repos_with_ids.txt
+.. branch is "8.0" for example
```
You can cleanup this comment

yvaucher commented on l10n_ch_lsv_dd/README.rst:
```diff
@@ -0,0 +1,106 @@
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/l10n-switzerland/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/l10n-switzerland/issues/new?body=module:%20l10n_ch_lsv_dd%0Aversion:%208.0.1.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
You can remove the link and just put a dot after `feedback`

yvaucher commented on l10n_ch_lsv_dd/README.rst:
```diff
@@ -0,0 +1,106 @@
+
+* Cyril Sester <cyril.sester@outlook.com>
+* Steve Ferry <steve.ferry1992@gmail.com>
+* brain-tec AG
```
Can we put individual names instead of company name ?

yvaucher commented on l10n_ch_lsv_dd/__init__.py:
```diff
@@ -0,0 +1,23 @@
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
You can remove this header from `__init__.py` file. Not much to put a license on here.

yvaucher commented on l10n_ch_lsv_dd/__manifest__.py:
```diff
@@ -0,0 +1,58 @@
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
Please change the headers to the new shorter format:

https://github.com/OCA/maintainer-tools/blob/master/template/module/models/model_name.py#L1-L3

yvaucher commented on l10n_ch_lsv_dd/__manifest__.py:
```diff
@@ -0,0 +1,58 @@
+    'category': 'Localization',
+    'depends': [
+        'l10n_ch_payment_slip',
+        'l10n_ch',
```
Is the default swiss chart of account mandatory or you can imagine not having l10n_ch in dependencies ?

yvaucher commented on l10n_ch_lsv_dd/models/__init__.py:
```diff
@@ -0,0 +1,26 @@
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
Header to remove

yvaucher commented on l10n_ch_lsv_dd/models/account_payment_method.py:
```diff
@@ -0,0 +1,50 @@
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
Short header to use here

yvaucher commented on l10n_ch_lsv_dd/models/account_payment_order.py:
```diff
@@ -0,0 +1,516 @@
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
short header to use

yvaucher commented on l10n_ch_lsv_dd/models/account_payment_order.py:
```diff
@@ -0,0 +1,516 @@
+from odoo import models, api, _
+import base64
+from lxml import etree
+from odoo.tools import float_round
```
It would recommand to order the imports this way:

``` python
import base64
from lxml import etree

from odoo import models, api, _
from odoo.tools import float_round
```


https://www.python.org/dev/peps/pep-0008/#imports

```
Imports should be grouped in the following order:

  1. standard library imports
  2. related third party imports
  3. local application/library specific imports
```

Ordering imports helps to have a better overview at first glance of what is used in the current file.

yvaucher commented on l10n_ch_lsv_dd/models/account_payment_order.py:
```diff
@@ -0,0 +1,516 @@
+    def show_invoices(self):
+        move_ids = [pay_line.move_line_id.move_id.id
+                    for pay_order in self
+                    for pay_line in pay_order.payment_line_ids]
```
Any reason why using this comprehension list instead of `mapped` ?

``` python
moves = self.mapped('payment_line_ids.move_line_id.move_id')
moves.ids
```

yvaucher commented on l10n_ch_lsv_dd/models/account_payment_order.py:
```diff
@@ -0,0 +1,516 @@
+        elif payment_method_code == 'sepa.ch.dd' and \
+                pain_flavor == 'pain.008.001.02.ch.03':
+            file_content, file_name = self.generate_xml_ch_dd_file()
+            res = file_content, file_name
```
s/file_content/file_contents

To be consistent with previous lines.

yvaucher commented on l10n_ch_lsv_dd/models/bank.py:
```diff
@@ -0,0 +1,119 @@
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
Short header to use.

yvaucher commented on l10n_ch_lsv_dd/models/bank.py:
```diff
@@ -0,0 +1,119 @@
+                if not self.is_post_dd_ident_valid(bank_account
+                                                   .post_dd_identifier):
+                    raise exceptions.ValidationError(_("Invalid Postfiance DD"
+                                                       "Identifier."))
```
Missing a space between "...DD" and "Identifier..."

yvaucher commented on l10n_ch_lsv_dd/models/bank.py:
```diff
@@ -0,0 +1,119 @@
+            lsv_identifier.decode('ascii')
+        except UnicodeDecodeError:
+            raise exceptions.ValidationError(
+                _('LSV identifier should contain only ASCII caracters.')
```
s/caracters/characters

yvaucher commented on l10n_ch_lsv_dd/security/ir.model.access.csv:
```diff
@@ -0,0 +1,3 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_ch_dd_export,Full access on banking.export.ch.dd,model_banking_export_ch_dd,sales_team.group_sale_manager,1,1,1,1
+access_ch_dd_export,Full access on banking.export.ch.dd,model_banking_export_ch_dd,sales_team.group_sale_salesman,1,1,1,1
```
The 2 rules have the same XMLID, thus only one will be in the system.

Please change the id and name of the second rule

yvaucher commented on l10n_ch_lsv_dd/tests/__init__.py:
```diff
@@ -0,0 +1,24 @@
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
Header can be remove

yvaucher commented on l10n_ch_lsv_dd/wizards/export_utils.py:
```diff
@@ -0,0 +1,125 @@
+    if not clean_ref == mod10r(clean_ref[0:26]):
+        return False
+
+    return True
```
As you depend from l10n_ch_payment_slip, don't you want to use the is_bvr_ref from there?

yvaucher commented on l10n_ch_lsv_dd/wizards/lsv_export_wizard.py:
```diff
@@ -0,0 +1,485 @@
+    237: 'i', 238: 'i', 239: 'i', 240: '.', 241: 'n', 242: 'o', 243: 'o',
+    244: 'o', 245: 'o', 246: 'eo', 247: '.', 248: '.', 249: 'u', 250: 'u',
+    251: 'u', 252: 'eu', 253: 'y', 254: '.', 255: 'y',
+}
```
Is it the same as in file l10n_ch_lsv_dd/wizards/export_utils.py ?

Why not importing it from there ?

yvaucher commented on l10n_ch_lsv_dd/wizards/lsv_export_wizard.py:
```diff
@@ -0,0 +1,485 @@
+    _description = 'Export LSV Direct Debit File'
+
+    treatment_type = fields.Selection(
+        [('P', _('Production')), ('T', _('Test'))],
```
When defining list of values directly in selection field without using a method you don't need to use `_()` as they will already be translatable.

yvaucher commented on l10n_ch_lsv_dd/wizards/lsv_export_wizard.py:
```diff
@@ -0,0 +1,485 @@
+        related='banking_export_ch_dd_id.total_amount'
+    )
+    state = fields.Selection(
+        [('create', _('Create')), ('finish', _('Finish'))],
```
Same here, you can remove the `_()`

yvaucher commented on l10n_ch_lsv_dd/wizards/lsv_export_wizard.py:
```diff
@@ -0,0 +1,485 @@
+            order_by += 'al.partner_bank_id'
+
+            # A direct db query is used because order parameter in model.search
+            # doesn't support function fields
```
order in `model.search` should be usable with stored computed fields. And by the way an SQL query will be as well impossible to be based on non stored fields.

Issue might be that you don't get a proper error as search will be silent about ordering by a non stored field.

ecino replied:
I did some corrections. However the code is not all mine, so I didn't change completely the SQL query. I just adapted it to the v10, because it was not working. I don't know if it's possible to use search with param such as oder='payment_line_ids.partner_bank_id' ? If it's ok for you I propose to keep this query (I removed the unused selected values).

yvaucher replied:
Ok we can keep it that way. But can you add a `# FIXME` to state that the current comment is not accurate?

ecino replied:
Done

yvaucher commented on l10n_ch_lsv_dd/wizards/lsv_export_wizard.py:
```diff
@@ -0,0 +1,485 @@
+            # A direct db query is used because order parameter in model.search
+            # doesn't support function fields
+            self.env.cr.execute("""
+                SELECT bl.id, m.date_maturity, al.partner_bank_id
```
While an SQL seems not required, why is this query including 3 columns while we use only the first one?

yvaucher commented on l10n_ch_lsv_dd/wizards/lsv_export_wizard.py:
```diff
@@ -0,0 +1,485 @@
+            payment_lines = payment_line_obj.browse(sorted_line_ids).exists()
+
+            for line in payment_lines:
+                if not line.mandate_id or not line.mandate_id.state == "valid":
```
`not ==` can be replaced by `!=`

And you can remove the test `line.mandate_id`

line will always be a recordset, and when you ask for a relation of a recordset, (even empty) you will get a record set.


```
>>> bool(line.mandate_id) == False
True
>>> line.mandate_id
account.banking.mandate()
>>> line.mandate_id.state
False
```

So getting an empty recordset will allow you to have your tests on the attributes of the relation without fearing it will crash.

yvaucher reviewed (changes requested):
Thanks for the porting here is a review on the whole PR

One bug in the ir.model.access.csv one question on SQL querries and mostly nitpickings.

yvaucher commented:
I don't know what is your CO-1571, I will squash you 2 commits into one.

Thanks for the porting, we can finally merge :)

yvaucher commented:
Merged at: https://github.com/OCA/l10n-switzerland/commit/76341dfaea743aafbcc878a095917a532afad04f
